extensions/interactions/base_test.py

```diff
@@ -246,6 +248,79 @@ def test_interaction_rule_descriptions_in_dict(self):
             'IsWithinTolerance': 'is within {{tol|Real}} of {{x|Real}}'
         })
 
+    def test_html_field_types_to_rule_specs_mapping_are_valid(self):
+        """Test that the structure of the file html_field_types_to_rule_specs.
+        json are valid. This test ensures that whenever any new type of
+        interaction or rule type with HTML string is added, the file
+        html_field_types_to_rule_specs.json should be updated accordingly.
+        """
+        # The file having the information about the assembly of the html in the
+        # rule specs.
+        html_field_types_to_rule_specs_dict = json.loads(
+            utils.get_file_contents(
+                feconf.HTML_FIELD_TYPES_TO_RULE_SPECS_FILE_PATH))
+
+        # The file having the templates for the structure of the rule specs.
+        # Contents of the file html_field_types_to_rule_specs.json will be
+        # verified against this file.
+        rule_descriptions_dict = json.loads(
+            utils.get_file_contents(feconf.RULES_DESCRIPTIONS_FILE_PATH))
+
+        # In the following part, we generate the html_field_types_to_rule_specs
+        # dict based on the values in the rule_descriptions.json file.
+        generated_html_field_types_dict = (
+            collections.defaultdict(lambda: collections.defaultdict(
+                lambda: collections.defaultdict(lambda: collections.defaultdict(
+                    lambda: collections.defaultdict(set))))))
+
+        # Verify that each html_type dict has a format key, which must be
+        # unique. After verification we delete the key for comparison with
+        # the generated html_field_types_to_rule_specs dict. We compare
+        # everything except the format, because the format can't be generated
+        # from the rule_descriptions.json file.
+        for html_type, html_type_dict in (
+                html_field_types_to_rule_specs_dict.items()):
+            self.assertTrue('format' in html_type_dict)
+            self.assertTrue(
+                html_type_dict['format'] in
+                feconf.ALLOWED_HTML_RULE_VARIABLE_FORMATS)
+            del html_type_dict['format']
+
+        html_types = []
```

Comment:
Also I'm kinda confused about why this is a "global". Why don't you just compute it on a per-description basis? After all, you only use it on a per-description basis, so wouldn't that involve less bookkeeping?

Reply:
done, changed